src/history/historyTracker.ts

```diff
@@ -420,13 +424,45 @@ export class HistoryTracker {
     return this.currentHistoryStep.marks;
   }
 
+  /**
+   * Returns true if we need to get the entire document's text
+   * to process an individual change
+   */
+  private _isDocumentTextNeeded(): boolean {
+    // Get the current document's text length.
+    const documentEnd = new Position(0, 0).getDocumentEnd();
+    const newTextLength = this.vimState.editor.document.offsetAt(documentEnd);
+
+    const isTextLengthDiff = this.oldText.length !== newTextLength;
+
+    // Determine if we just switched modes.
+    // This prevents recording steps in between start-end of a historyStep.
+    const isModeDiff = this.currentMode !== this.vimState.currentMode;
+
+    const isNewHistoryStep =
+      (this.currentHistoryStepIndex === this.historySteps.length - 1 &&
+        this.currentHistoryStep.isFinished) ||
+      this.currentHistoryStepIndex !== this.historySteps.length - 1;
+
+    if (isModeDiff) {
+      this.currentMode = this.vimState.currentMode;
+    }
+
+    // If all of these are false we can avoid requesting the entire doc.
+    return (isTextLengthDiff && (isNewHistoryStep || isModeDiff)) || isNewHistoryStep || isModeDiff;
```

Comment:
This seems wrong or at least overcomplicated. Isn't this equivalent to just `isNewHistoryStep || isModeDiff`?

Reply:
Yes, you're right!. Thanks for catching it, it completely went over my head. Originally I wanted to just check the text length to avoid getting the whole doc when just moving around, but that doesn't cover cases where an actual change doesn't affect the length of the text. In trying to solve that, I think I got my own logic mixed up, sorry about that.

Reply:
Checking the text length is  not a viable solution, I'm going to remove that.